File: api/src/grpc/post.rs
use core_services::services::base::Service;
use schema::devlog::{
    devblog::rpc::{
        post_interaction_request,
        post_interaction_response::InteractionResult,
        post_service_server::PostService as PostGrpcService,
        CreatePostRequest,
        CreatePostResponse,
        GetPostRequest,
        GetPostResponse,
        PostInteractionRequest,
        PostInteractionResponse
    },
    entities::{Like, User}
};
use tonic::{Request, Response, Status};

use crate::{di::ApiDependenciesInjection, services::post::*};

#[derive(Clone)]
pub struct PostGrpcServer {
    pub(crate) di: &'static ApiDependenciesInjection
}

#[async_trait::async_trait]
impl PostGrpcService for PostGrpcServer {
    async fn get(&self, request: Request<GetPostRequest>) -> Result<Response<GetPostResponse>, tonic::Status> {
        let request = request.get_ref();
        let service = self.di.get_post_service();

        let params = GetPostParams {
            title: request.title.clone()
        };

        let result = service.execute(params).await?;
        Ok(Response::new(GetPostResponse {
            total_likes: result.total_likes,
            total_views: result.total_views,
            post: Some(result.post)
        }))
    }

    async fn create(&self, request: Request<CreatePostRequest>) -> Result<Response<CreatePostResponse>, tonic::Status> {
        let user: &User = request.extensions().get::<User>().ok_or(Status::unauthenticated("You're not authorize"))?;

        let service = self.di.create_post_service();

        let request = request.get_ref();
        let params = CreatePostParams {
            post: request.post.clone().unwrap(),
            user: user.clone()
        };

        let result = service.execute(params).await?;
        Ok(Response::new(CreatePostResponse {
            post: Some(result.post)
        }))
    }

    async fn interact(&self, request: Request<PostInteractionRequest>) -> Result<Response<PostInteractionResponse>, tonic::Status> {
        let user: &User = request.extensions().get::<User>().ok_or(Status::unauthenticated("You're not authorize"))?;

        let service = self.di.interaction_service();

        let request = request.get_ref();
        let post_id = request.id.as_ref().expect("The post id is required").clone();
        let interaction = match &request.interaction {
            Some(post_interaction_request::Interaction::Like(like)) => Interaction::Like(like.clone()),
            Some(post_interaction_request::Interaction::Vote(_vote)) => Interaction::Vote,
            Some(post_interaction_request::Interaction::View(_view)) => Interaction::View,
            _ => return Err(Status::invalid_argument("Not support this interaction in post")),
        };

        let like_post_params = PostInteractionParams {
            user: user.clone(),
            post_id,
            interaction
        };

        let result = service.execute(like_post_params).await?;
        match result {
            PostInteractionResult::Like(count)=> Ok(Response::new(PostInteractionResponse{interaction_result:Some(InteractionResult::TotalLikeCount(count))})),
            PostInteractionResult::View(count) => Ok(Response::new(PostInteractionResponse{interaction_result:Some(InteractionResult::TotalViewCount(count))})),
            PostInteractionResult::Vote(count) => Ok(Response::new(PostInteractionResponse{interaction_result:Some(InteractionResult::TotalVoteCount(count))})),
        }
    }
}

File: web/src/app/services/post.ts
import {
  PostId,
  Like,
  View,
  Vote,
  PostService as PostServiceClient,
  PostInteractionRequest,
  GetPostRequest,
  Post, GetPostResponse
} from "@devlog/schema-ts"
import gRPCClientBase from "./base"
import PostLocalStorage from "../storage/post"

export default class PostService extends gRPCClientBase<typeof PostServiceClient> {
  postStorage: PostLocalStorage

  constructor(postStorage: PostLocalStorage) {
    super(PostServiceClient)
    this.postStorage = postStorage
  }

  async get(postTitle: string): Promise<{ post: Post, totalLikes: number, totalViews: number }> {
    const request = new GetPostRequest()
    request.title = postTitle

    const response = await this.client.get(request, { headers: this.getHeader() }) as GetPostResponse

    return {
      post: response.post!,
      totalLikes: response.totalLikes,
      totalViews: response.totalViews
    }
  }

  async view(postTitle: string): Promise<number> {
    return new Promise((resolve, reject) => {
      setTimeout(async () => {
        const postTitleWords = postTitle.toLowerCase().split(' ')
        const url = window.location.href.toLowerCase()
        const isInTheSameWebsite = postTitleWords.every(word => url.includes(word))

        if (isInTheSameWebsite) {
          const request = new PostInteractionRequest()
          const postId = new PostId()
          postId.title = postTitle
          request.id = postId
          request.interaction = {value: new View(), case: 'view'}

          const response = await this.client.interact(request)
          resolve(response.interactionResult.value || 0)
        } else {
          resolve(0)
        }
      }, 10000)
    })
  }

  async like(postTitle: string, count: number): Promise<number> {
    const request = new PostInteractionRequest()
    const postId = new PostId()
    postId.title = postTitle
    request.id = postId

    const like = new Like()
    like.count = count
    request.interaction = {value: like, case: 'like'}

    const response = await this.client.interact(request, { headers: this.getHeader() })
    return response.interactionResult.value || 0
  }

  async vote(postTitle: string): Promise<number> {
    const request = new PostInteractionRequest()
    const postId = new PostId()
    postId.title = postTitle
    request.id = postId
    request.interaction = {value: new Vote(), case: 'vote'}

    const response = await this.client.interact(request, { headers: this.getHeader() })
    await this.postStorage.addVotedPost(postTitle)
    return response.interactionResult.value || 0
  }
}